### PR TITLE
[Active Record] Support row-constructor clauses for PostgreSQL/MySQL

### PR DESCRIPTION
**Pull request by nvasilevski**
### Motivation / Background

This PR addresses the https://github.com/rails/rails/issues/40273 issue and shares similarities with https://github.com/rails/rails/pull/36003

Previously opened issues and pull requests with similar proposals show the demand and use-cases for querying records by tuples (sets of column values) in bulk. Most of database engines allow rows to be queried “in tuples” with the following (or similar) syntax: `select * from my_table where (column1, column2) IN ((value1, value2,))`. 

Introduction of `query_constraints` in https://github.com/rails/rails/pull/46331 allows for a single instance of an Active Record model to be updated/reloaded/destroyed using a set of two or more columns in the query. Having a model that is being accessed or even identified by a set of columns eventually leads to a need to query the Model objects in bulk. Which creates additional demand for the “querying by tuples” feature being available in Rails.

This PR extends `.where` relation builder API to support row-constructor clauses.

#### General approach
The current implementation diverts from the existing flow as early as possible, right in the `build_where_clause` call and explicitly builds a newly introduced `Arel::Nodes::RowConstructor` node. This has a huge advantage of not needing to either integrate the `RowConstructor` into the whole query building flow either by coming up with a somewhat complex abstraction or just adding more branching like `attrs.is_a?(Array)` and `attrs.is_a?(AttributeSet)` to accommodate new "attributes set" concept. However it also means we are losing some capabilities that the existing logic has. The most obvious that comes to mind - aliases support `klass.attribute_aliases[key] || key`. At the same time it is easier to maintain the new feature as from the beginning it won't be overcomplicated with capabilities while still serving most of the use-cases.
Introduction of two new classes - `AttributesSet` and `RowConstructor`

#### AttributesSet
`AttributesSet` at this point is a much more simplified version of the `Arel` `Attribute` which will represent a combination of more than one attribute in sql context.
There was an alternative to the `AttributesSet` - to fit the "multiple attribute names" concept into the current `Attribute` class which would either mean a big refactoring or several uncomfortable `if` conditions and weird naming where `#name` method returns an Array of values instead of a single value.

#### RowConstructor

`RowConstructor` - new arel node, almost a complete copy of the existing `HomogeneousIn` node Added specifically to work with the `AttributesSet` instead of `Attribute`. Similar to the `AttributesSet`, there was an alternative to change existing `HomogeneousIn` to be able to work with `AttributesSet` but it would have resulted in an `if` condition. Having a separate `RowConstructor` node is a better option for extensibility.

### Currently available alternatives to the proposed feature:

1. Manually built sql - `ClothingItem.where("(clothing_type, color) IN ((?, ?), (?, ?))", *["pants", "green", "tshirt", "yellow"]))`. Far from being convenient as requires manual `(?, ?)` template assembly 
2. Series of `OR` statements, for example: `where (clothing_type=pants and clothing_color=green) OR (clothing_type="tshirt" and clothing_color=yellow)` doesn't scale - not the most readable code when applying multiple `.or` statements and the resulting SQL is larger than with the row-constructor syntax. 

So basically there is no decent alternative to the proposed feature when it comes to larger number of records to be loaded.

### Previous concerns

The main previous concern was focused on the previously proposed API which required the row-constructor clause to be manually built. We are proposing a higher level-api that allows `.where` to accept a hash like `Article.where({ [:title, :text] => [['title1', 'text1'], ['title2', 'text2']] })` which removes the need of a manually built clause. Potentially we can go even higher and introduce a separate helper like `Article.row_constructor(columns, values).to_a` if extending `.where` API doesn’t seem suitable.

### Row-constructor name

We are naming this feature in the same way it is being called in MySQL guides - https://dev.mysql.com/doc/refman/8.0/en/row-subqueries.html

### Supported adapters

We are starting from supporting Postgres and MySQL database engines as both of those support the same `WHERE (c1, c2) IN ((v1, v2), (v3, v4)) ` syntax. Eventually the support may be extended to SQLite with a slightly different syntax: `(a, b) IN (VALUES (5, 6), (2, 4))` as shown in the table https://github.com/rails/rails/issues/40273#issuecomment-709017516

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters/abstract/quoting.rb [hunk 1]
```diff
@@ -54,6 +55,10 @@ def quote_bound_value(value)
         quote(value)
       end
 
+      def quote_array(a)
```
Comment (eileencodes) [inline activerecord/lib/active_record/connection_adapters/abstract/quoting.rb:58]: In don't think we want this to be part of the public API so we can move it under private or at least add nodoc.

File: activerecord/lib/active_record/relation/query_methods.rb [hunk 2]
```diff
@@ -1396,16 +1396,27 @@ def build_where_clause(opts, rest = []) # :nodoc:
...
             key = key.to_s
             klass.attribute_aliases[key] || key
           end
           references = PredicateBuilder.references(opts)
           self.references_values |= references unless references.empty?
 
-          parts = predicate_builder.build_from_hash(opts) do |table_name|
+          parts_from_builder = predicate_builder.build_from_hash(opts) do |table_name|
             lookup_table_klass_from_join_dependencies(table_name)
           end
```
Comment (eileencodes) [inline activerecord/lib/active_record/relation/query_methods.rb:1418]:
```suggestion
          end
          
```

**Review by eileencodes — COMMENTED** [review]
I requested review from @matthewd since he was originally against the first iteration of the feature but now that we have a more solid use case we should get his opinion. I also requested review from @tenderlove for the Arel bits.

I left a couple comments about making methods private.

**Comment by matthewd** [comment]
I think this still introduces the array-quoting behaviour I pushed back on in #36003, encoding that the correct way to send an array value to Postgres is `(1,2)` and not the actual array literal `ARRAY[1,2]`. Let's not do that.

Bypassing alias resolution is also IMO not okay, as well as seemingly unnecessary. It's useful to take small steps in introducing new behaviour, but to land, it needs to fit in with the existing functionality, not introduce an "either or" trade-off between different feature sets. I see it happening in `Table#[]` -- can we not just do it in the `map` as well?

RowConstructor is a bad name for the node as it is currently defined. `(1, 2)` is a row constructor. `(1, 2) IN ((1, 2), (3, 4))` is not -- it's an IN expression (which contains row constructors).

That being said, this feels like a rather oversized node shape anyway. Can we represent this using a regular `In` node with children for the contained row constructors? You mentioned this is modelled on `HomogeneousIn`, but that's a bad outlier, which sacrifices fundamental Arel design principles for performance on a very hot codepath. Starting from there feels like we're putting the cart before the horse. If we do end up going that way, from a quick glance, I think the two can be easily unified by having `HomogeneousIn` accept either an `Attribute` or `AttributeSet`, and replacing the conditionals you mentioned with polymorphism.

Finally, if `.or` creates awkward queries for relatively simple inputs, can we just fix it? Even if we add this node _and_ the proposed public API (which IMO is a little clunky, but an elegant solution given the constraints), if "that makes a bad query" is a problem, and better options are within reach, we should make it not do that?

**Comment by benedikt** [comment]
What about introducing a `Arel::Nodes::Tuple` type to model all kinds of `(1, 2, 3)` constructs instead? 

Not sure about the performance implications, but it would make a useful building block in general. 

In a way, `ValuesList` and `HomogeneousIn` and `RowConstructor` all end up building this SQL fragment plus whatever else they need for their specific type. 

Things like `((1, 2), (3, 4))` can be a tuple of tuples. And by keeping the node type generic, attributes and aliases can be handled in child nodes.

**Comment by nvasilevski** [comment]
Closing this PR in favor of https://github.com/rails/rails/pull/47729

Since we have decoupled the `.where([] => [])` syntax from the row-constructor SQL this PR will need to be reworked when we decide to extend the API with the row constructor sql

**Pull request closed without merging**